### PR TITLE
fix: don't replace fragment on another fragment if they are the same

### LINKED ISSUE #3465: The app seems to duplicate queries
The app seems to duplicate queries
```
XXX.92 - - [23/Aug/2020:11:40:03 +0200] "GET /1.json?fields=null HTTP/2.0" 200 98979 "-" "Open Food Facts Official Android App 3.4.2"
XXX.92 - - [23/Aug/2020:11:40:03 +0200] "GET /files/tagline/tagline-off.json HTTP/2.0" 200 2164 "-" "Open Food Facts Official Android App 3.4.2"
XXX.92 - - [23/Aug/2020:11:40:03 +0200] "GET /files/tagline/tagline-off.json HTTP/2.0" 200 2164 "-" "Open Food Facts Official Android App 3.4.2"
XXX.92 - - [23/Aug/2020:11:40:03 +0200] "GET /1.json?fields=null HTTP/2.0" 200 98979 "-" "Open Food Facts Official Android App 3.4.2"
```

### PR DESCRIPTION
**Pull request by naivekook**
####  Description
when user clicks on "Home" button twice we did a replacement of current fragment on new instance of this fragment and do extra calls to the server

#### Related issues
Fixes #3465

### CODE REVIEW

**Comment by naivekook** [comment]
@VaiTon @teolemon please take a look

**Comment by VaiTon** [comment]
I don't really like this implementation using tag string. Can't we compare using `is`, for example? Is there a way to get the actual displayed fragment?

**Comment by naivekook** [comment]
> I don't really like this implementation using tag string.

Can I ask you why?

fragment manager doesn't have convenient mechanism to retrieve current fragment, by combining `findFragmentByTag` and `fragment.isVisible` we can understand this

looks like we use `replace` everywhere without backstack so we have only one fragment attached, in this case I can do next thing and it will work too
```
with((currentActivity as FragmentActivity).supportFragmentManager) {
                        val fragment = fragments.lastOrNull()
                        if (fragment == null || fragment !is HomeFragment) {
                            commit {
                                replace(R.id.fragment_container, HomeFragment.newInstance(), HomeFragment.TAG)
                                addToBackStack(null)
                            }
                        }
                    }
```

**Comment by VaiTon** [comment]
> 
> 
> > I don't really like this implementation using tag string.
> 
> Can I ask you why?

Just because it's a useless var if we can check the type as you showed IMHO. 

About your second solution, if you could test it and see if it works it would be awesome!

**Comment by naivekook** [comment]
@VaiTon hey check updated version please
I've been tested it on emulator and it works

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/MainActivity.kt [hunk 1]
```diff
@@ -422,9 +417,17 @@ class MainActivity : BaseActivity(), NavigationDrawerListener {
         handleIntent(intent)
     }
 
-    private fun swapToHomeFragment() {
-        supportFragmentManager.addOnBackStackChangedListener {}
-        supportFragmentManager.commit { replace(R.id.fragment_container, HomeFragment()) }
+    private fun swapToFragment(fragment: Fragment) {
+        with(supportFragmentManager) {
```
Comment (VaiTon) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/features/MainActivity.kt]: Really hate saying this but i think `let` would be more clear when calling methods. Could you replace `with` with `let`?

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/MainActivity.kt [hunk 2]
```diff
@@ -422,9 +417,17 @@ class MainActivity : BaseActivity(), NavigationDrawerListener {
...
+            val currentFragment = fragments.lastOrNull()
+            if (currentFragment == null || currentFragment::class.java != fragment::class.java) {
```
Comment (VaiTon) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/features/MainActivity.kt]: Why don't we use `currentFragment !is fragment::class`?

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/searchbycode/SearchByCodeFragment.kt [hunk 3]
```diff
@@ -76,10 +76,10 @@ class SearchByCodeFragment : NavigationBaseFragment() {
 
     companion object {
         const val INTENT_KEY_BARCODE = "barcode"
-        fun newInstance(barcode: String) = SearchByCodeFragment().apply {
+        fun newInstance(barcode: String? = null) = SearchByCodeFragment().apply {
```
Comment (VaiTon) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/features/searchbycode/SearchByCodeFragment.kt:79]: Why this? We cannot search by code without any code.

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/listeners/CommonBottomListener.kt [hunk 4]
```diff
@@ -34,9 +34,14 @@ class CommonBottomListener internal constructor(private val currentActivity: Act
             }
             R.id.home_page, R.id.home -> {
                 if (isCurrentActivity(WelcomeActivity::class.java) || isCurrentActivity(MainActivity::class.java)) {
-                    (currentActivity as FragmentActivity).supportFragmentManager.commit {
-                        replace(R.id.fragment_container, HomeFragment())
-                        addToBackStack(null)
+                    with((currentActivity as FragmentActivity).supportFragmentManager) {
```
Comment (VaiTon) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/features/listeners/CommonBottomListener.kt]: Could we extract a method out of this?

**Review by VaiTon — CHANGES_REQUESTED** [review]

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/searchbycode/SearchByCodeFragment.kt (hunk 3 above)
Comment (naivekook) [reply app/src/main/java/openfoodfacts/github/scrachx/openfood/features/searchbycode/SearchByCodeFragment.kt:79]: sorry I don't get you

actually I see that we dont have any usages of this method so we can remove it safely I guess

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/MainActivity.kt (hunk 2 above)
Comment (naivekook) [reply app/src/main/java/openfoodfacts/github/scrachx/openfood/features/MainActivity.kt]: your variant isn't working

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/MainActivity.kt (hunk 1 above)
Comment (naivekook) [reply app/src/main/java/openfoodfacts/github/scrachx/openfood/features/MainActivity.kt]: well this is more "style dispute", removed `with`

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/listeners/CommonBottomListener.kt (hunk 4 above)
Comment (naivekook) [reply app/src/main/java/openfoodfacts/github/scrachx/openfood/features/listeners/CommonBottomListener.kt]: sure, done

**Comment by naivekook** [comment]
ptal @VaiTon

**Review by VaiTon — APPROVED** [review]

**Comment by VaiTon** [comment]
Thanks! 🎉

**Pull request merged**